File: python/examples/02_filter_by_exchange.py
"""Filter the snapshot by a single exchange and print the first 10 rows sorted by base."""

from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from kairo_funding import FundingClient, cli_entry  # noqa: E402


def main() -> int:
    exchange = os.environ.get("KAIRO_EXCHANGE", "bybit")
    with FundingClient() as client:
        snapshot = client.get_snapshot(exchange=exchange)
    rows = FundingClient.parse_compact_rows(snapshot.get("data", []))
    count = snapshot.get("count", len(rows))
    print(f"exchange={exchange}  rows={count}")
    rows.sort(key=lambda r: r["base"])
    for row in rows[:10]:
        print(
            f"{row['base']}  rate={row['funding_rate']:.6f}  "
            f"interval={row['funding_interval_hours']}h"
        )
    return 0


if __name__ == "__main__":
    cli_entry(main)

File: typescript/src/examples/02_filter_by_exchange.ts
/**
 * 02_filter_by_exchange: demonstrates the `exchange` query filter and
 * prints the first 10 rows sorted by base in a column-aligned table.
 */

import { FundingClient } from "../client.js";
import { cliEntry } from "../runner.js";

function signed(value: number, decimals: number): string {
  return (value >= 0 ? "+" : "") + value.toFixed(decimals);
}

async function main(): Promise<void> {
  const exchange = process.env.KAIRO_EXCHANGE ?? "bybit";
  const client = new FundingClient();
  const snap = await client.getSnapshot({ exchange });

  process.stdout.write(`exchange=${exchange}  rows=${snap.count}\n`);
  process.stdout.write("\n");
  process.stdout.write(
    `${"base".padEnd(16)}  ${"rate".padStart(11)}  ${"intv".padStart(4)}\n`,
  );

  const sorted = [...snap.data].sort((a, b) => a.base.localeCompare(b.base));
  for (const row of sorted.slice(0, 10)) {
    process.stdout.write(
      `${row.base.padEnd(16)}  ` +
        `${signed(row.fundingRate, 6).padStart(11)}  ` +
        `${String(row.fundingIntervalHours).padStart(3)}h\n`,
    );
  }
}

cliEntry(main);
